subprojects/plugins/src/main/java/org/gradle/api/tasks/GroovyRuntime.java

```diff
@@ -127,25 +128,22 @@ private FileCollection inferGroovyClasspath() {
 
                 VersionNumber groovyVersion = groovyJar.getVersion();
 
-                // Groovy 3 does not have groovy-all yet we may have the required pieces on classpath via localGroovy()
-                if (groovyVersion.getMajor() == 3) {
-                    return inferGroovy3Classpath(groovyVersion);
+                // We may already have the required pieces on classpath via localGroovy()
+                if (groovyVersion.equals(VersionNumber.parse(GroovySystem.getVersion()))) {
```

Comment:
Looking at the test and the old production code closer, I think that this change is wrong: For any Groovy version >= 3, `inferClasspath` should use the files already on the classpath, even for versions different than the Groovy version. This is not only for `localGroovy`, but also when your configuration already has the necessary dependencies. The Groovy version is encoded in the JAR names, so the check would fail if the other jars are from a different version.